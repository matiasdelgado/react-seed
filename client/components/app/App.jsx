import React from 'react';
import PropTypes from 'prop-types';
import Button from '../button';
import Counter from '../counter';
import classes from 'utils/classes';
import styles from './styles.scss';

class App extends React.Component {
	constructor(props) {
		super(props);
		this.handleClick = this.handleClick.bind(this);
	}

	render() {
		const { loading, samples, error } = this.props.app;
		const text = (loading && 'loading...') || error;
		const samplesArray = samples.slice();

		return (
			<div className={classes(styles.container, styles.shadow)}>
				<Counter />
				<button className={classes(styles.button, styles.shadow)} onClick={this.handleClick}>
					Test
				</button>
				<div>
					{!!text && <Button text={text} />}
					{!text &&
						samplesArray.map(s =>
							<div key={s.id}>
								<Button text={s.description} />
							</div>
						)}
				</div>
			</div>
		);
	}

	handleClick() {
		this.props.app.fetchSamples();
	}
}

App.propTypes = {
	app: PropTypes.shape({
		error: PropTypes.string,
		fetchSamples: PropTypes.func,
		loading: PropTypes.bool,
		samples: PropTypes.object
	})
};

export default App;
