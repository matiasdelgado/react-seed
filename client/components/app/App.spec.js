import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import { observable } from 'mobx';
import App from './App';

const samples = observable([
	{ id: 1, description: 'sample #1' },
	{ id: 2, description: 'sample #2' },
	{ id: 3, description: 'sample #3' }
]);

const appState = {
	error: '',
	fetchSamples: jest.fn(),
	loading: false,
	samples
};

const PROPS = { app: appState };

describe('<App/>', () => {
	const renderer = new ShallowRenderer();

	it('render no samples, no loading and no error', () => {
		const tree = renderer.render(<App {...PROPS} />);

		expect(tree).toMatchSnapshot();
	});

	it('render samples', () => {
		const tree = renderer.render(<App {...PROPS} samples={samples} />);

		expect(tree).toMatchSnapshot();
	});

	it('render loading...', () => {
		const tree = renderer.render(<App {...PROPS} loading />);

		expect(tree).toMatchSnapshot();
	});

	it('render error', () => {
		const tree = renderer.render(<App {...PROPS} error="something went wrong!" />);

		expect(tree).toMatchSnapshot();
	});

	it('render loading before samples', () => {
		const tree = renderer.render(<App {...PROPS} samples={samples} loading />);

		expect(tree).toMatchSnapshot();
	});

	it('render error before samples', () => {
		const tree = renderer.render(<App {...PROPS} samples={samples} error="error!" />);

		expect(tree).toMatchSnapshot();
	});
});

test('click on button should call props.fetchSamples func', () => {
	const app = shallow(<App {...PROPS} />);

	app.find('button').simulate('click');

	expect(PROPS.app.fetchSamples).toBeCalled();
});
