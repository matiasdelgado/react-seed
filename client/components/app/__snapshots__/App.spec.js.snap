// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> render error 1`] = `
<div
  className="container shadow"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
  <div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #1"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #2"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #3"
      />
    </div>
  </div>
</div>
`;

exports[`<App/> render error before samples 1`] = `
<div
  className="container shadow"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
  <div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #1"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #2"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #3"
      />
    </div>
  </div>
</div>
`;

exports[`<App/> render loading before samples 1`] = `
<div
  className="container shadow"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
  <div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #1"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #2"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #3"
      />
    </div>
  </div>
</div>
`;

exports[`<App/> render loading... 1`] = `
<div
  className="container shadow"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
  <div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #1"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #2"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #3"
      />
    </div>
  </div>
</div>
`;

exports[`<App/> render no samples, no loading and no error 1`] = `
<div
  className="container shadow"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
  <div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #1"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #2"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #3"
      />
    </div>
  </div>
</div>
`;

exports[`<App/> render samples 1`] = `
<div
  className="container shadow"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
  <div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #1"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #2"
      />
    </div>
    <div>
      <Button
        onClick={[Function]}
        text="sample #3"
      />
    </div>
  </div>
</div>
`;
